### PR TITLE
Saveless Updates

### LINKED ISSUE #306: Rearchitect signify to diff the buffer
*Note: This is work I think I have time to do, I just want to ensure it is welcome before starting.*

# Objective

There are a number of features that I would like to see in signify which don't seem easy/possible with the current VCS interface. I would like to adjust this slightly to support these feautres.

- Updates without saving buffer.
- [revert hunks](https://github.com/mhinz/vim-signify/issues/119)

Furthermore this will give us full control over the diff which should make parsing the diff simpler and allow for smarter diffing algorithms. This should transitively improve [diff preview](https://github.com/mhinz/vim-signify/issues/300) because of more flexible diffing (example, showing more context or using vim's diffing support).

Other posibilities unlcoked:
- Diff against any other file on the filesystem.
- Diff against the version of the file when opened.
- DIff against the file on disk. (unsaved changes)
- Cache base version for slow VCS (avoid hitting a remote server too often).

# Proposal

## VCS Interface

The proposal is quite simple, instead of asking the VCS to perform the diff we will just ask it to dump the "base" version of the file. This changes the VCS interface from diff to dump and will require reimplementing all of the current VCS commands. However since this interface is quite simple this should be easy to do. Furthermore we can support falling back to the old diff command until we add support for all VCS using the new system.

Example implementations:

```sh
git show :$file
hg cat $file
```

## Getting the Diff

The diff will be obtained by dumping both versions to disk and running the diff. Note that this doesn't require saving the buffer over top of the canonical verison of the file.

Additionally we now run the diff consistently across VCSs which gives us a lot of posibilities, including running different types of diffs for different things (for example diffing with context for previewing the change but without context for generating signs).

## Concerns

- Diffing locally can be faster than diffing remotely. There is the posibility for massive files transfering the diff from the server is faster. However this would require uploading the local version to the server so it is unlikely this would have better performance. Additionally this design opens the posibility of caching the base version so it will probably be faster if anything.
- Not all VCS support dumping the base version. This is possible but I have yet to see one. After all their whole job is storing revisions.
- The code would be hard to modify to implement this new interface. I haven't looked at the code so I have no idea.
- This would break user configs. If users have implemented custom diff commands then we would prefer these for a while. Eventually we may wish to deprecate support for this. Users will have to migration to the new API if they want the new features anyways.

### PR DESCRIPTION
**Pull request by kevincox**
This updates signify to diff against the buffer instead of the on-disk file. This for example makes `:SignifyRefresh` work properly.

This will allow us to remove autosave from realtime mode.

Fixes #306, #230

### CODE REVIEW

File: autoload/sy/repo.vim [hunk 1]
```diff
@@ -342,20 +413,32 @@ endfunction
 " Function: s:initialize_job {{{1
 function! s:initialize_job(vcs) abort
   let vcs_cmd = s:expand_cmd(a:vcs, g:signify_vcs_cmds)
+  return s:wrap_cmd(a:vcs, vcs_cmd)
+endfunction
+
+" Function: s:initialize_buffer_job {{{1
+function! s:initialize_buffer_job(vcs, live) abort
+  let base_cmd = s:get_base_cmd(a:vcs)
+  let diff_cmd = 'set -o pipefail; ' . base_cmd . ' | ' . s:difftool . ' -U0 - ' . fnameescape(a:live)
```
Comment (mhinz) [inline autoload/sy/repo.vim]: From a first glance, `set -o pipefail` probably won't work in shells like fish, Powershell, or cmd.exe.

Comment (kevincox) [reply autoload/sy/repo.vim]: I misunderstood thinking that it was always wrapped in `sh -c` but I now see that is wrong. I don't think there is a portable way to do this however I can just implement using a temp file. The performance difference should be negligible.

**Comment by kevincox** [comment]
If you are happy with the feature and general implementation direction I will clean up the request.

- [ ] Use portable shell syntax only.
- [ ] Move behind preference.
- [ ] Rebase.
- [ ] Allow choosing either implementation so that "undo" feature can always use the live buffer.

**Comment by mhinz** [comment]
Yeah, I like this feature a lot. :-)

We probably need to go the temporary file route to avoid compatibility issues with Windows etc. (So, no shell-specific stuff, not even pipes.)

> Allow choosing either implementation so that "undo" feature can always use the live buffer.

You mean keeping the current way _and_ the new, saveless one? I guess this would be useful. Many users prefer updating signs on saving only, thus on `BufWritePost`. In that case, using a temporary file would be unnecessary.

This feature is the last thing missing for vim-signify 2.0. When it's implemented, [a lot of old options can be removed](https://github.com/mhinz/vim-signify/blob/13ba32ae556b1d6e4e354f5647f6f1ab91d10e55/plugin/signify.vim#L25-L59). Then the plugin will act as if `g:signify_realtime` was always on. (For users who dislike updating on moving the cursor, the doc will tell them how to remove the autocommands for that.)

**Comment by kevincox** [comment]
On Sat 5 Oct 2019, 20:24 Marco Hinz, <notifications@github.com> wrote:

> Yeah, I like this feature a lot. :-)
>
Great 🙂

> You mean keeping the current way *and* the new, saveless one? I guess
> this would be useful. Many users prefer updating signs on saving only, thus
> on BufWritePost. In that case, using a temporary file would be
> unnecessary.
>
Yes, at least initially I would implement it as opt-in for initial testing.
(Possibly always on for the undo feature since a: it is more reliable and
2: it won't beark anyone because it is a new feature)

> This feature is the last thing missing for vim-signify 2.0. When it's
> implemented, a lot of old options can be removed
> <https://github.com/mhinz/vim-signify/blob/13ba32ae556b1d6e4e354f5647f6f1ab91d10e55/plugin/signify.vim#L25-L59>.
> Then the plugin will act as if g:signify_realtime was always on. (For
> users who dislike updating on moving the cursor, the doc will tell them how
> to remove the autocommands for that.)
>
That sounds good. I do think we should roll it after beta testing this for
a while first though.

>

**Comment by mhinz** [comment]
Hiding it behind an option sounds good to me. Let's just get this into master and work from there. 

Oh, and ignore documentation for now. I'll do this when this PR is merged (together with `:SignifyHunkPreview` and `:SignifyHunkUndo`.)

**Comment by kevincox** [comment]
Okay, everything is looking pretty good now.

- Behind `g:signify_live` preference.
  - All diffing will use the buffer if this is set.
  - `:SignifyHunkUndo` always uses the live diff.
- Uses what I believe is portable sh syntax.

One major question: Right now the "live" diff automatically falls back to the old code if the VCS isn't supported. Should we keep this? It can be confusing for users if they set `g:signify_live` but don't get live diffing. This will also lead us to getting support for the rest of the VCSs quicker.

**Comment by mhinz** [comment]
> Only 2 VCS don't support the live diff anyways so it is probably better to give an error for them.

I just added the 2 missing commands. I actually tested the RCS one and for AccuRev.. well, it _should_ work according to the [CLI user guide](https://supportline.microfocus.com/Documentation/books/AccuRev/AccuRev/6.2/accurev-620-cliusersguide-en.pdf), page 58.

So, no falling back needed anymore.

**Comment by kevincox** [comment]
Great. Removing the fallback is trivial (just remove half of the condition in the `if`) I should get around to that on Monday. Other than that I think it is ready to go so please make a final review.

File: autoload/sy/repo.vim [hunk 2]
```diff
@@ -365,20 +435,41 @@ endfunction
...
+  return s:wrap_cmd(a:vcs, vcs_cmd)
+endfunction
+
+" Function: s:initialize_buffer_job {{{1
+function! s:initialize_buffer_job(vcs) abort
+  let bufferfile = tempname()
+  call s:write_buffer(bufferfile)
+
+  let basefile = tempname()
+  let base_cmd = s:get_base_cmd(a:vcs) . '>' . fnameescape(basefile) . ' && '
```
Comment (mhinz) [inline autoload/sy/repo.vim]: I think this should work with cmd.exe, but fail with Powershell.. but that doesn't matter for now. If there's no other way, I'll simply force `cmd.exe` on Windows (like we do with `sh` on all other systems).

**Comment by mhinz** [comment]
Let's just merge this and work from there! I'll do a few cosmetic changes in a follow-up commit.

Thanks for your work!  ✨

**Pull request merged**